#162948847 Favorite and Unfavorite Articles

KMaina opened the pull request:
#### What does this PR do?
Allows for users to favorite and unfavorite articles
#### Description of Task to be completed?
This endpoint favorites and unfavorites articles depending on the state
`PUT api/articles/<slug:slug>/favorite`
#### How should this be manually tested?

- Clone the repo locally
- cd ah-django-unchained
- git checkout ft-favorite-unfavorite-articles-162948847
- get started with the app as from the README section. Follow the sample .env file supplied to gert further started
- create an article as required.
- once you hit the `PUT api/articles/<slug:slug>/favorite` endpoint, the favorite status should be `True`. this should also apply to the `GET api/articles/:slug` endpoint
<img width="1134" alt="favorite article" src="https://user-images.githubusercontent.com/31338414/51696942-c13f5280-2017-11e9-9c9c-a4e2eec2e37e.png">

- to unfavorite hit the `PUT api/articles/<slug:slug>/favorite` endpoint, and check the same in the `GET api/articles/:slug` endpoint
<img width="1126" alt="favorite article false" src="https://user-images.githubusercontent.com/31338414/51697042-f055c400-2017-11e9-8684-85c00970d9be.png">

#### What are the relevant pivotal tracker stories?
[#162948847](https://www.pivotaltracker.com/story/show/162948847)

KMaina committed:
feat(favorite articles): add migrations from the models

[Delivers #162948847]

KMaina committed:
feat(favorite articles): add models for favoriting articles

- add a favorite field in the model

[Delivers #162948847]

KMaina committed:
feat(favorite articles): add favorite field in the serializer

[Delivers #162948847]

KMaina committed:
feat(favorite articles): write unit tests for favoriting articles

[Delivers #162948847]

KMaina committed:
feat(favorite articles): add views for favoriting and unfavoriting an article

[Delivers #162948847]

KMaina committed:
feat(favorite articles): add the url that handles favoriting

[Delivers #162948847]

kwameasiago commented on authors/apps/articles/serializers.py:
```diff
@@ -1,7 +1,7 @@
 
 from rest_framework import serializers
-from .models import Article
+from .models import Article #, FavoriteArticle
```
Remove commented code

kwameasiago commented on authors/apps/articles/serializers.py:
```diff
@@ -62,3 +66,19 @@ class DeleteArticleSerializer(serializers.ModelSerializer):
         fields = ['is_deleted']
+
+
+# class FavoriteSerializer(serializers.ModelSerializer):
```
Remove commented code

kwameasiago commented on authors/apps/articles/urls.py:
```diff
@@ -1,14 +1,17 @@
 
-from .views import ArticleAPIView, ArticleDetailsView, DeleteArticle
+from .views import ArticleAPIView, ArticleDetailsView, DeleteArticle, FaveArticle
 
```
Add an extra line between the imports and app_name

Georgeygigz reviewed (approved):
Looks good

kwameasiago reviewed (approved):
Looks good to me

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -11,6 +11,8 @@ def setUp(self):
         self.login_url = reverse('authentication:auth-login')
         self.signup_url = reverse('authentication:auth-register')
         self.article_listcreate = reverse('articles:articles-listcreate')
+        self.article_delete = reverse('articles:articles-favorite',
```
Why `article_delete` as variable name? It's not an appropriate name to hold the favourite url

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -198,3 +200,35 @@ def test_to_fetch_non_exisiting_article(self):
+                         self.create_article_data,
+                         format='json',
+                         HTTP_AUTHORIZATION='token {}'.format(token))
+        self.client.put(reverse('articles:articles-favorite',
```
I think you should use the variable created to hold the url since you declared it above

ignatiusukwuoma commented on authors/apps/articles/urls.py:
```diff
@@ -1,14 +1,19 @@
+     path('articles/delete/<slug:slug>/',
+          DeleteArticle.as_view(),
+          name="articles-delete"),
+     path('articles/<slug:slug>/favorite',
```
WHy do we use `<slug:slug>` instead of just `<slug>`?

KMaina replied:
Changed

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -50,9 +51,60 @@ class ArticleDetailsView(RetrieveUpdateAPIView):
+            # get an article that matches the slug specified
+            # and return a message if the article does not exist
+            article = Article.objects.get(slug=slug)
+        except ObjectDoesNotExist:
```
Did you import `ObjectDoesNotExist`?

KMaina replied:
I have imported

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -50,9 +51,60 @@ class ArticleDetailsView(RetrieveUpdateAPIView):
+            # and return a message if the article does not exist
+            article = Article.objects.get(slug=slug)
+        except ObjectDoesNotExist:
+            raise NotFound("This article does not exist")
```
Did you import `NotFound`?

KMaina replied:
I have imported

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -50,9 +51,60 @@ class ArticleDetailsView(RetrieveUpdateAPIView):
+        # if it is favorited we unfavorite it
+        if article in Article.objects.filter(
+                                        favorite=request.user):
+            article.favorite.remove(request.user)
```
Let us have a `DELETE` request handle unfavoriting articles according to the specifications provided

KMaina replied:
Done

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -50,9 +51,60 @@ class ArticleDetailsView(RetrieveUpdateAPIView):
+
+def set_favorite_status(data, user):
+        """Set the favorite value to either True or False"""
+        # Check if the user has favorited the article
```
I thought you checked this in the view above?

KMaina replied:
This function checks if a user has favorited or unfavorited an article, then instead of displaying the id we set it to True or False

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -50,9 +51,60 @@ class ArticleDetailsView(RetrieveUpdateAPIView):
+        # Check if the user has favorited the article
+        value = [value for value in data.data['favorite'] if
+                 user in data.data['favorite']]
+        # Article has not been favorited, set favorite to False
```
Article has been favorited you mean?

KMaina replied:
I have refactored the code and changes this

ignatiusukwuoma commented:
Rebase your branch too

KMaina committed:
feat(favorite articles): refactor tests for favoriting articles

- add more tests to test additional implementation
- refactor current tests for new functionality

[Delivers #162948847]

KMaina committed:
feat(favorite articles): change favoriting implementation

- add a delete method to unfavorite articles
- add a message if an article is already favorited or unfavorited

[Delivers #162948847]

KMaina committed:
feat(favorite articles): rename the urls for the app

- have the slugs as <slug>

[Delivers #162948847]

KMaina committed:
feat(favorite articles): merge migration files

[Delivers #162948847]

KMaina committed:
feat(favorite articles): rebase from master and merge migration files

[Delivers #162948847]

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -206,3 +208,71 @@ def test_to_fetch_non_exisiting_article(self):
+        """Test to favorite an article"""
+        token = self.signup_user_one()
+        del self.create_article_data['images']
+        self.client.post(self.article_listcreate,
```
I'm sure you don't want us to focus on this part.. so create a helper function that creates an article so you can call it in one line and we can focus on the favorite endpoint below.

KMaina replied:
Done, I used the helper functions elsewhere to reduce duplication

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -206,3 +208,71 @@ def test_to_fetch_non_exisiting_article(self):
+        self.client.post(self.article_favorite,
+                         format='json',
+                         HTTP_AUTHORIZATION='token {}'.format(token))
+        response = self.client.delete(self.article_favorite,
```
Make some assertions above this to confirm that article has been favorited

KMaina replied:
Done, I also added the same some assertions for unfavoriting an article

KMaina committed:
feat(favorite articles): refactor tests

- add assertions
- remove duplication

[Delivers #162948847]
